### PR TITLE
Fixes #23227, #24623, #24626, #18951 - Ansible Roles API

### PR DESCRIPTION
**Pull request by ofedoren**
Cherry-picked and rebased from [PR 161](https://github.com/theforeman/foreman_ansible/pull/161), [PR 198](https://github.com/theforeman/foreman_ansible/pull/198), [PR 199](https://github.com/theforeman/foreman_ansible/pull/199), [PR 200](https://github.com/theforeman/foreman_ansible/pull/200).

P.S. [Hammer plugin](https://github.com/theforeman/hammer-cli-foreman-ansible) currently doesn't have commands for these changes/endpoints, but I've prepared them and could create a PR there with them for testing. Anyway I was going to do it, but after this is merged (for a case of additional changes in here).

### CODE REVIEW

File: app/controllers/foreman_ansible/api/v2/hosts_controller_extensions.rb [hunk 1]
```diff
@@ -6,31 +6,63 @@ module HostsControllerExtensions
...
+          api :GET, '/hosts/:id/ansible_roles',
+              N_('List all Ansible roles for a host')
+          param :id, :identifier, :required => true
+
+          def ansible_roles
+            return unless @host
+            @ansible_roles = @host.all_ansible_roles
+          end
+
+          api :POST, '/hosts/:id/assign_ansible_roles',
```
Comment (dLobatog) [inline app/controllers/foreman_ansible/api/v2/hosts_controller_extensions.rb]: Why this endpoint? It seems to me that we can definitely do this already by sending a POST request to `api/v2/hosts/:id` with the 'ansible_roles' attribute

File: app/controllers/foreman_ansible/api/v2/hostgroups_controller_extensions.rb [hunk 2]
```diff
@@ -6,31 +6,59 @@ module HostgroupsControllerExtensions
...
+              N_('List all Ansible roles for a hostgroup')
+          param :id, :identifier, :required => true
+
+          def ansible_roles
+            find_resource
+            return unless @hostgroup
+            @ansible_roles = @hostgroup.all_ansible_roles
+          end
+
+          api :POST, '/hostgroups/:id/assign_ansible_roles',
```
Comment (dLobatog) [inline app/controllers/foreman_ansible/api/v2/hostgroups_controller_extensions.rb]: Same question as for the one for this endpoint in hosts.

File: app/controllers/api/v2/ansible_roles_controller.rb [hunk 3]
```diff
@@ -41,8 +42,27 @@ def obsolete
         @obsoleted = @importer.obsolete!
       end
 
+      api :GET, '/ansible_roles/fetch',
```
Comment (dLobatog) [inline app/controllers/api/v2/ansible_roles_controller.rb]: Nice job!

File: app/controllers/foreman_ansible/concerns/job_invocation_helper.rb [hunk 4]
```diff
@@ -7,10 +7,14 @@ module JobInvocationHelper
       def job_composer(feature_name, target)
         composer = ::JobInvocationComposer.for_feature(feature_name, target)
         return composer if composer.save
-        raise ::Foreman::Exception.new(
-          format(N_('Could not run Ansible roles for %{host}'),
-                 :host => target)
-        )
+        msg = if target.blank?
+                format(N_('There are no Ansible roles to play'),
```
Comment (dLobatog) [inline app/controllers/foreman_ansible/concerns/job_invocation_helper.rb]: Why passing `:target => target.class` to format, if it's not used at all?

File: app/controllers/foreman_ansible/api/v2/hosts_controller_extensions.rb (hunk 1 above)
Comment (ofedoren) [reply app/controllers/foreman_ansible/api/v2/hosts_controller_extensions.rb]: Sure we can, It's just added "from" https://github.com/theforeman/foreman_ansible/pull/81/files#diff-e3a172920758b4851cabcb23ced54a22R64 and also for easier use from hammer. If you think this is redundant, tell me and I'll remove that.

**Comment by ofedoren** [comment]
@dLobatog ping. I rebased, but tests are red even though 0 skipped, 0 failed?

**Comment by ares** [comment]
@ofedoren we discussed this today with bigger group, it would be great to send a separate PR with assign_* endpoints. Sorry for back and forth. I'm merging this meanwhile, test failure is only on sqlite because of some machine problem.

**Pull request merged**